### PR TITLE
Emit events via CLI

### PR DESCRIPTION
**Pull request by aduane**
This adds a CLI command to emit an event to a locally running plugin-runner based on fixture data.

The CLI will pick a random event from the provided file and send it to the plugin runner. The user is able to specify the port of the plugin runner optionally, but it defaults to `50051`.

Some fixture files are provided, and should be expanded on, but we can do that on an ongoing basis. I wanted to get the functionality out. I will create a PR for the docs site to document this CLI functionality once/if this PR is approved.

Expected fixture file format:
---
Newline-delimited JSON, each line a single object containing the attributes: `EventType`, `target`, and `context`.
If you wanted to create a fixture, you could write a plugin that subscribes to all events (`RESPONDS_TO = EventType.keys()`, but you almost certainly wouldn't want to do that in a production environment) and logs these values.

Example `ASSESS_COMMAND__CONDITION_SELECTED.ndjson`

```json
{"EventType": "ASSESS_COMMAND__CONDITION_SELECTED", "target": "72943f6d-fa4b-49e7-b92b-64975b516d0a", "context": "{\"fields\": {\"condition\": {\"text\": \"Bitten by orca, initial encounter (W56.21XA)\", \"disabled\": false, \"description\": null, \"annotations\": null, \"extra\": null, \"value\": 25}, \"background\": \"\", \"status\": \"\", \"narrative\": \"\"}, \"note\": {\"uuid\": \"0c81daca-cf89-4f9b-bc3d-6f1ae0644f27\"}, \"data\": {\"condition\": {\"codings\": [{\"code\": \"404684003\", \"system\": \"http://snomed.info/sct\", \"display\": \"Bitten by orca, initial encounter\"}, {\"code\": \"W5621XA\", \"system\": \"ICD-10\", \"display\": \"Bitten by orca, initial encounter\"}], \"externally_exposable_id\": \"6dfcc4b4-1ede-4d6d-a89a-1c847fecb9f0\"}}}"}
```

### CODE REVIEW

File: canvas_cli/main.py [hunk 1]
```diff
@@ -1,12 +1,19 @@
 import importlib.metadata
+import json
+import random
 from pathlib import Path
 from typing import Optional
 
+import grpc
 import typer
+from typing_extensions import Annotated
```
Comment (csande) [inline canvas_cli/main.py:9]: I think `typing_extensions` might befor backward compatibility, but `Annotated` lives in `typing` after 3.11.

```suggestion
from typing import Annotated
```

File: canvas_cli/main.py [hunk 2]
```diff
@@ -51,6 +58,37 @@ def get_or_create_config_file() -> Path:
...
+    ] = "50051",
+) -> None:
+    """
+    Grab an event from a fixture file and send it your locally running plugin-runner process.
+    Any resultant effects will be printed.
+
+    Valid fixture files are newline-delimited JSON, with each containing the keys `EventType`, `target`, and `context`. Some fixture files are included in the canvas-plugins repo.
+    """
+    # Grab a random event from the fixture file ndjson
+    lines = open(event_fixture).read().splitlines()
```
Comment (csande) [inline canvas_cli/main.py:75]: I think that syntax leaves the file handle open.

```suggestion
    with open(event_fixture) as file_:
        lines = file_.read().splitlines()
```

Comment (beaugunderson) [reply canvas_cli/main.py:75]: you can do it without a context if you use pathlib, e.g. `Path(event_fixture).read_all().splitlines()`

File: canvas_cli/main.py [hunk 3]
```diff
@@ -1,12 +1,19 @@
...
 from typing import Optional
 
+import grpc
 import typer
+from typing_extensions import Annotated
 
 from canvas_cli.apps import plugin
 from canvas_cli.apps.logs import logs as logs_command
 from canvas_cli.utils.context import context
+from generated.messages.events_pb2 import Event as PluginRunnerEvent
```
Comment (beaugunderson) [inline canvas_cli/main.py:14]: some linter should grumble about this, imports from the same file should be grouped

File: canvas_cli/main.py [hunk 4]
```diff
@@ -51,6 +58,37 @@ def get_or_create_config_file() -> Path:
...
+    """
+    Grab an event from a fixture file and send it your locally running plugin-runner process.
+    Any resultant effects will be printed.
+
+    Valid fixture files are newline-delimited JSON, with each containing the keys `EventType`, `target`, and `context`. Some fixture files are included in the canvas-plugins repo.
+    """
+    # Grab a random event from the fixture file ndjson
+    lines = open(event_fixture).read().splitlines()
+    myline = random.choice(lines)
+    event_data = json.loads(myline)
```
Comment (jrwils) [inline canvas_cli/main.py:77]: Why do we want to make this random? If I am a plugin developer and my plugin has a specific event that it responds to, I think there would be more value in being able to specify the event that is sent.

File: canvas_cli/main.py (hunk 3 above)
Comment (aduane) [reply canvas_cli/main.py:14]: Strangely, isort is causing this. I group them, isort separates.

Screenshot 2024-06-17 at 3 13 37 PM

Comment (beaugunderson) [reply canvas_cli/main.py:14]: oh, well, I will not argue with `isort`! ignore me haha

File: canvas_cli/main.py (hunk 2 above)
Comment (aduane) [reply canvas_cli/main.py]: Looks like it's `Path(event_fixture).read_text().splitlines()`

File: canvas_cli/main.py (hunk 4 above)
Comment (aduane) [reply canvas_cli/main.py:76]: Similar to randomizing the test order in pytest, I can see value in ensuring your plugin behaves as expected in a variety of scenarios. If a developer wants to test against a fixed event, they are able to create their own ndjson with a single line and pass that file path in as the CLI arg.

**Review by beaugunderson — APPROVED** [review]

**Pull request merged**